compiler/semtypes.nim

```diff
@@ -1021,26 +1029,42 @@ proc semProcTypeNode(c: PContext, n, genericParams: PNode,
       length = sonsLen(a)
       hasType = a.sons[length-2].kind != nkEmpty
       hasDefault = a.sons[length-1].kind != nkEmpty
+
     if hasType:
       typ = semParamType(c, a.sons[length-2], constraint)
 
     if hasDefault:
-      def = semExprWithType(c, a.sons[length-1])
-      # check type compatibility between def.typ and typ:
+      def = a[^1]
+      block determineType:
+        if genericParams != nil and genericParams.len > 0:
+          def = semGenericStmt(c, def)
+          if hasUnresolvedArgs(c, def):
+            def.typ = makeTypeFromExpr(c, def.copyTree)
+            break determineType
+
+        def = semExprWithType(c, def, {efDetermineType})
+
       if typ == nil:
         typ = def.typ
-      elif def != nil:
-        # and def.typ != nil and def.typ.kind != tyNone:
+        if typ.kind == tyTypeDesc:
+          # default typedesc values are mapped to the unbound typedesc type:
```

Comment:
Ok, I'll bite. Where in the spec is this mentioned?

Reply:
This should be obvious. When you create a proc such as

```nim
proc takesType(T = typedesc[int])
```

It would be extremely surprising if you cannot pass the `string` type to the proc. The rule above ensures that the compiler sees the definition above as `proc takesType(T: type)`. I don't think I need to explain this in the manual, because it's completely common-sense.

Reply:
Yes. Guess what, you need to explain it better here, in the compiler's source code.